File: src/example_plugin.py
"""
Example Synapse Plugin Implementation

This module demonstrates a complete plugin implementation with:
- Tool registration and execution
- Configuration management
- Lifecycle hooks
- State management
- Error handling
- Async operations
"""

import asyncio
import logging
from typing import Any, Dict, List, Optional
from datetime import datetime

try:
    import requests
    from pydantic import BaseModel, Field, validator
except ImportError:
    # Fallback for environments without dependencies
    requests = None
    BaseModel = object
    Field = lambda **kwargs: None
    validator = lambda *args, **kwargs: lambda f: f


class PluginConfig(BaseModel):
    """Plugin configuration schema with validation"""
    
    api_key: str = Field(default="", description="API key for external service")
    endpoint: str = Field(default="https://api.example.com", description="API endpoint URL")
    timeout: int = Field(default=30, ge=1, le=300, description="Request timeout in seconds")
    max_retries: int = Field(default=3, ge=0, le=10, description="Maximum retry attempts")
    enable_caching: bool = Field(default=True, description="Enable response caching")
    log_level: str = Field(default="INFO", description="Logging level")
    
    @validator("log_level")
    def validate_log_level(cls, v):
        valid_levels = ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
        if v.upper() not in valid_levels:
            raise ValueError(f"log_level must be one of {valid_levels}")
        return v.upper()
    
    @validator("endpoint")
    def validate_endpoint(cls, v):
        if not v.startswith(("http://", "https://")):
            raise ValueError("endpoint must start with http:// or https://")
        return v


class ExamplePlugin:
    """
    Example Synapse Plugin
    
    Demonstrates core plugin functionality including tool execution,
    lifecycle management, state handling, and error recovery.
    """
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        """
        Initialize the plugin
        
        Args:
            config: Plugin configuration dictionary
        """
        self.config = PluginConfig(**(config or {}))
        self.logger = self._setup_logger()
        self.state: Dict[str, Any] = {
            "initialized_at": datetime.utcnow().isoformat(),
            "execution_count": 0,
            "cache": {} if self.config.enable_caching else None,
            "errors": []
        }
        self.session_data: Dict[str, Any] = {}
        
        self.logger.info(f"ExamplePlugin initialized with config: {self.config.dict()}")
    
    def _setup_logger(self) -> logging.Logger:
        """Configure plugin logger"""
        logger = logging.getLogger(f"synapse.plugin.{self.__class__.__name__}")
        logger.setLevel(getattr(logging, self.config.log_level))
        
        if not logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
            handler.setFormatter(formatter)
            logger.addHandler(handler)
        
        return logger
    
    # -------------------------------------------------------------------------
    # Lifecycle Hooks
    # -------------------------------------------------------------------------
    
    async def on_startup(self) -> None:
        """Called when the plugin is loaded"""
        self.logger.info("Plugin starting up...")
        
        # Perform startup tasks
        if self.config.api_key:
            self.logger.info("API key configured")
        
        # Validate external dependencies
        if requests is None:
            self.logger.warning("requests library not available")
        
        self.logger.info("Plugin startup complete")
    
    async def on_shutdown(self) -> None:
        """Called when the plugin is being unloaded"""
        self.logger.info("Plugin shutting down...")
        
        # Cleanup resources
        if self.state.get("cache"):
            cache_size = len(self.state["cache"])
            self.logger.info(f"Clearing cache with {cache_size} entries")
            self.state["cache"].clear()
        
        # Log statistics
        self.logger.info(
            f"Total executions: {self.state['execution_count']}, "
            f"Errors: {len(self.state['errors'])}"
        )
        
        self.logger.info("Plugin shutdown complete")
    
    async def on_session_start(self, session_id: str, context: Dict[str, Any]) -> None:
        """
        Called when a new session starts
        
        Args:
            session_id: Unique session identifier
            context: Session context data
        """
        self.logger.info(f"Session started: {session_id}")
        self.session_data[session_id] = {
            "started_at": datetime.utcnow().isoformat(),
            "context": context,
            "tool_calls": []
        }
    
    async def on_session_end(self, session_id: str) -> None:
        """
        Called when a session ends
        
        Args:
            session_id: Session identifier
        """
        if session_id in self.session_data:
            session_info = self.session_data[session_id]
            tool_calls = len(session_info.get("tool_calls", []))
            self.logger.info(
                f"Session ended: {session_id}, Tool calls: {tool_calls}"
            )
            del self.session_data[session_id]
    
    async def on_tool_execute(
        self, 
        tool_name: str, 
        parameters: Dict[str, Any],
        session_id: Optional[str] = None
    ) -> None:
        """
        Called before a tool is executed
        
        Args:
            tool_name: Name of the tool being executed
            parameters: Tool parameters
            session_id: Associated session ID
        """
        self.logger.debug(f"Executing tool: {tool_name} with params: {parameters}")
        
        if session_id and session_id in self.session_data:
            self.session_data[session_id]["tool_calls"].append({
                "tool": tool_name,
                "timestamp": datetime.utcnow().isoformat(),
                "parameters": parameters
            })
    
    async def on_error(
        self, 
        error: Exception, 
        context: Dict[str, Any]
    ) -> None:
        """
        Called when an error occurs
        
        Args:
            error: The exception that occurred
            context: Error context information
        """
        error_info = {
            "timestamp": datetime.utcnow().isoformat(),
            "error": str(error),
            "type": type(error).__name__,
            "context": context
        }
        
        self.state["errors"].append(error_info)
        self.logger.error(f"Error occurred: {error}", exc_info=True)
    
    # -------------------------------------------------------------------------
    # Tool Implementations
    # -------------------------------------------------------------------------
    
    async def example_tool(
        self,
        query: str,
        options: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Example tool that processes a query
        
        Args:
            query: Input query to process
            options: Optional processing options
                - format: Output format (json, text, xml)
                - verbose: Enable verbose output
        
        Returns:
            Processing results
        """
        self.logger.info(f"example_tool called with query: {query}")
        self.state["execution_count"] += 1
        
        options = options or {}
        output_format = options.get("format", "json")
        verbose = options.get("verbose", False)
        
        # Check cache
        cache_key = f"{query}:{output_format}:{verbose}"
        if self.config.enable_caching and cache_key in self.state.get("cache", {}):
            self.logger.debug(f"Cache hit for: {cache_key}")
            return self.state["cache"][cache_key]
        
        # Process query
        result = {
            "status": "success",
            "query": query,
            "format": output_format,
            "timestamp": datetime.utcnow().isoformat(),
            "processed": query.upper() if output_format == "text" else {
                "original": query,
                "length": len(query),
                "words": len(query.split())
            }
        }
        
        if verbose:
            result["metadata"] = {
                "execution_count": self.state["execution_count"],
                "cache_enabled": self.config.enable_caching,
                "plugin_version": "1.0.0"
            }
        
        # Update cache
        if self.config.enable_caching:
            self.state["cache"][cache_key] = result
        
        return result
    
    async def fetch_data(
        self,
        resource_id: str,
        fields: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        Fetch data from external API
        
        Args:
            resource_id: Resource identifier to fetch
            fields: Specific fields to retrieve
        
        Returns:
            Fetched data
        """
        self.logger.info(f"fetch_data called for resource: {resource_id}")
        self.state["execution_count"] += 1
        
        if not self.config.api_key:
            self.logger.warning("No API key configured, returning mock data")
            return {
                "status": "mock",
                "resource_id": resource_id,
                "fields": fields or [],
                "data": {"message": "Mock data - configure api_key for real data"}
            }
        
        if requests is None:
            raise RuntimeError("requests library not available")
        
        # Construct API request
        url = f"{self.config.endpoint}/resources/{resource_id}"
        headers = {"Authorization": f"Bearer {self.config.api_key}"}
        params = {}
        
        if fields:
            params["fields"] = ",".join(fields)
        
        # Execute request with retry logic
        for attempt in range(self.config.max_retries + 1):
            try:
                self.logger.debug(f"API request attempt {attempt + 1}/{self.config.max_retries + 1}")
                
                response = requests.get(
                    url,
                    headers=headers,
                    params=params,
                    timeout=self.config.timeout
                )
                response.raise_for_status()
                
                return {
                    "status": "success",
                    "resource_id": resource_id,
                    "data": response.json()
                }
            
            except requests.RequestException as e:
                self.logger.warning(f"Request failed (attempt {attempt + 1}): {e}")
                
                if attempt < self.config.max_retries:
                    await asyncio.sleep(2 ** attempt)  # Exponential backoff
                else:
                    raise
        
        raise RuntimeError("Max retries exceeded")
    
    # -------------------------------------------------------------------------
    # Utility Methods
    # -------------------------------------------------------------------------
    
    def get_state(self) -> Dict[str, Any]:
        """
        Get current plugin state
        
        Returns:
            Plugin state dictionary
        """
        return {
            "config": self.config.dict(),
            "state": {
                **self.state,
                "cache": f"{len(self.state.get('cache', {}))} entries" if self.state.get("cache") else "disabled"
            },
            "sessions": len(self.session_data)
        }
    
    def clear_cache(self) -> None:
        """Clear the plugin cache"""
        if self.state.get("cache"):
            self.state["cache"].clear()
            self.logger.info("Cache cleared")
    
    def get_metrics(self) -> Dict[str, Any]:
        """
        Get plugin metrics
        
        Returns:
            Metrics dictionary
        """
        return {
            "execution_count": self.state["execution_count"],
            "error_count": len(self.state["errors"]),
            "cache_size": len(self.state.get("cache", {})),
            "active_sessions": len(self.session_data),
            "uptime": (
                datetime.utcnow() - 
                datetime.fromisoformat(self.state["initialized_at"])
            ).total_seconds()
        }


# Export plugin class
__all__ = ["ExamplePlugin"]

File: src/main/java/dev/synapse/plugin/example/ExamplePlugin.java
package dev.synapse.plugin.example;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Example Synapse Plugin Implementation
 * 
 * This class demonstrates a complete plugin implementation with:
 * - Tool registration and execution
 * - Configuration management
 * - Lifecycle hooks
 * - State management
 * - Error handling
 * - Async operations
 */
@Slf4j
@Component
public class ExamplePlugin {
    
    private final PluginConfig config;
    private final PluginState state;
    private final Map<String, SessionData> sessionData;
    
    public ExamplePlugin(PluginConfig config) {
        this.config = config;
        this.state = new PluginState();
        this.sessionData = new ConcurrentHashMap<>();
        
        log.info("ExamplePlugin initialized with config: {}", config);
    }
    
    // -------------------------------------------------------------------------
    // Lifecycle Hooks
    // -------------------------------------------------------------------------
    
    /**
     * Called when the plugin is loaded
     */
    public void onStartup() {
        log.info("Plugin starting up...");
        
        // Perform startup tasks
        if (config.getApiKey() != null && !config.getApiKey().isEmpty()) {
            log.info("API key configured");
        }
        
        log.info("Plugin startup complete");
    }
    
    /**
     * Called when the plugin is being unloaded
     */
    public void onShutdown() {
        log.info("Plugin shutting down...");
        
        // Cleanup resources
        if (config.isEnableCaching() && state.getCache() != null) {
            int cacheSize = state.getCache().size();
            log.info("Clearing cache with {} entries", cacheSize);
            state.getCache().clear();
        }
        
        // Log statistics
        log.info("Total executions: {}, Errors: {}", 
            state.getExecutionCount().get(), 
            state.getErrors().size());
        
        log.info("Plugin shutdown complete");
    }
    
    /**
     * Called when a new session starts
     * 
     * @param sessionId Unique session identifier
     * @param context Session context data
     */
    public void onSessionStart(String sessionId, Map<String, Object> context) {
        log.info("Session started: {}", sessionId);
        SessionData data = new SessionData(sessionId, context);
        sessionData.put(sessionId, data);
    }
    
    /**
     * Called when a session ends
     * 
     * @param sessionId Session identifier
     */
    public void onSessionEnd(String sessionId) {
        SessionData data = sessionData.get(sessionId);
        if (data != null) {
            int toolCalls = data.getToolCalls().size();
            log.info("Session ended: {}, Tool calls: {}", sessionId, toolCalls);
            sessionData.remove(sessionId);
        }
    }
    
    /**
     * Called before a tool is executed
     * 
     * @param toolName Name of the tool being executed
     * @param parameters Tool parameters
     * @param sessionId Associated session ID
     */
    public void onToolExecute(String toolName, Map<String, Object> parameters, String sessionId) {
        log.debug("Executing tool: {} with params: {}", toolName, parameters);
        
        if (sessionId != null) {
            SessionData data = sessionData.get(sessionId);
            if (data != null) {
                data.addToolCall(toolName, parameters);
            }
        }
    }
    
    /**
     * Called when an error occurs
     * 
     * @param error The exception that occurred
     * @param context Error context information
     */
    public void onError(Exception error, Map<String, Object> context) {
        ErrorInfo errorInfo = new ErrorInfo(error, context);
        state.getErrors().add(errorInfo);
        log.error("Error occurred: {}", error.getMessage(), error);
    }
    
    // -------------------------------------------------------------------------
    // Tool Implementations
    // -------------------------------------------------------------------------
    
    /**
     * Example tool that processes a query
     * 
     * @param request Tool request containing query and options
     * @return Processing results
     */
    public Map<String, Object> exampleTool(ExampleToolRequest request) {
        log.info("example_tool called with query: {}", request.getQuery());
        state.getExecutionCount().incrementAndGet();
        
        String outputFormat = request.getOptions() != null ? 
            request.getOptions().getOrDefault("format", "json").toString() : "json";
        boolean verbose = request.getOptions() != null && 
            Boolean.TRUE.equals(request.getOptions().get("verbose"));
        
        // Check cache
        String cacheKey = String.format("%s:%s:%s", request.getQuery(), outputFormat, verbose);
        if (config.isEnableCaching() && state.getCache().containsKey(cacheKey)) {
            log.debug("Cache hit for: {}", cacheKey);
            return state.getCache().get(cacheKey);
        }
        
        // Process query
        Map<String, Object> result = new HashMap<>();
        result.put("status", "success");
        result.put("query", request.getQuery());
        result.put("format", outputFormat);
        result.put("timestamp", Instant.now().toString());
        
        if ("text".equals(outputFormat)) {
            result.put("processed", request.getQuery().toUpperCase());
        } else {
            Map<String, Object> processed = new HashMap<>();
            processed.put("original", request.getQuery());
            processed.put("length", request.getQuery().length());
            processed.put("words", request.getQuery().split("\\s+").length);
            result.put("processed", processed);
        }
        
        if (verbose) {
            Map<String, Object> metadata = new HashMap<>();
            metadata.put("execution_count", state.getExecutionCount().get());
            metadata.put("cache_enabled", config.isEnableCaching());
            metadata.put("plugin_version", "1.0.0");
            result.put("metadata", metadata);
        }
        
        // Update cache
        if (config.isEnableCaching()) {
            state.getCache().put(cacheKey, result);
        }
        
        return result;
    }
    
    /**
     * Fetch data from external API
     * 
     * @param request Fetch data request
     * @return Fetched data
     */
    public Map<String, Object> fetchData(FetchDataRequest request) {
        log.info("fetch_data called for resource: {}", request.getResourceId());
        state.getExecutionCount().incrementAndGet();
        
        if (config.getApiKey() == null || config.getApiKey().isEmpty()) {
            log.warn("No API key configured, returning mock data");
            Map<String, Object> result = new HashMap<>();
            result.put("status", "mock");
            result.put("resource_id", request.getResourceId());
            result.put("fields", request.getFields() != null ? request.getFields() : Collections.emptyList());
            result.put("data", Map.of("message", "Mock data - configure api_key for real data"));
            return result;
        }
        
        // In a real implementation, use WebClient or RestTemplate to make API calls
        // This is a placeholder showing the structure
        Map<String, Object> result = new HashMap<>();
        result.put("status", "success");
        result.put("resource_id", request.getResourceId());
        result.put("data", Map.of("message", "Real API call would go here"));
        
        return result;
    }
    
    // -------------------------------------------------------------------------
    // Utility Methods
    // -------------------------------------------------------------------------
    
    /**
     * Get current plugin state
     * 
     * @return Plugin state information
     */
    public Map<String, Object> getState() {
        Map<String, Object> stateInfo = new HashMap<>();
        stateInfo.put("config", config);
        
        Map<String, Object> stateData = new HashMap<>();
        stateData.put("initialized_at", state.getInitializedAt());
        stateData.put("execution_count", state.getExecutionCount().get());
        stateData.put("cache", config.isEnableCaching() ? 
            state.getCache().size() + " entries" : "disabled");
        stateData.put("errors", state.getErrors().size());
        stateInfo.put("state", stateData);
        
        stateInfo.put("sessions", sessionData.size());
        
        return stateInfo;
    }
    
    /**
     * Clear the plugin cache
     */
    public void clearCache() {
        if (state.getCache() != null) {
            state.getCache().clear();
            log.info("Cache cleared");
        }
    }
    
    /**
     * Get plugin metrics
     * 
     * @return Metrics information
     */
    public Map<String, Object> getMetrics() {
        Map<String, Object> metrics = new HashMap<>();
        metrics.put("execution_count", state.getExecutionCount().get());
        metrics.put("error_count", state.getErrors().size());
        metrics.put("cache_size", state.getCache().size());
        metrics.put("active_sessions", sessionData.size());
        
        long uptime = Instant.now().getEpochSecond() - state.getInitializedAt().getEpochSecond();
        metrics.put("uptime", uptime);
        
        return metrics;
    }
    
    // -------------------------------------------------------------------------
    // Inner Classes
    // -------------------------------------------------------------------------
    
    /**
     * Plugin state management
     */
    private static class PluginState {
        private final Instant initializedAt = Instant.now();
        private final AtomicLong executionCount = new AtomicLong(0);
        private final Map<String, Map<String, Object>> cache = new ConcurrentHashMap<>();
        private final List<ErrorInfo> errors = Collections.synchronizedList(new ArrayList<>());
        
        public Instant getInitializedAt() { return initializedAt; }
        public AtomicLong getExecutionCount() { return executionCount; }
        public Map<String, Map<String, Object>> getCache() { return cache; }
        public List<ErrorInfo> getErrors() { return errors; }
    }
    
    /**
     * Session data tracking
     */
    private static class SessionData {
        private final String sessionId;
        private final Instant startedAt = Instant.now();
        private final Map<String, Object> context;
        private final List<Map<String, Object>> toolCalls = Collections.synchronizedList(new ArrayList<>());
        
        public SessionData(String sessionId, Map<String, Object> context) {
            this.sessionId = sessionId;
            this.context = context;
        }
        
        public void addToolCall(String toolName, Map<String, Object> parameters) {
            Map<String, Object> call = new HashMap<>();
            call.put("tool", toolName);
            call.put("timestamp", Instant.now().toString());
            call.put("parameters", parameters);
            toolCalls.add(call);
        }
        
        public List<Map<String, Object>> getToolCalls() { return toolCalls; }
    }
    
    /**
     * Error information tracking
     */
    private static class ErrorInfo {
        private final Instant timestamp = Instant.now();
        private final String error;
        private final String type;
        private final Map<String, Object> context;
        
        public ErrorInfo(Exception exception, Map<String, Object> context) {
            this.error = exception.getMessage();
            this.type = exception.getClass().getSimpleName();
            this.context = context;
        }
    }
}
